include/rocksdb/trace_record_result.h

```diff
@@ -0,0 +1,153 @@
+ public:
+  TraceExecutionResult(uint64_t start_timestamp, uint64_t end_timestamp,
+                       TraceType trace_type);
+
+  // Execution start/end timestamps and request latency in microseconds.
+  virtual uint64_t GetStartTimestamp() const;
+  virtual uint64_t GetEndTimestamp() const;
+  inline uint64_t GetLatency() const {
+    return GetEndTimestamp() - GetStartTimestamp();
+  }
+
+ private:
+  uint64_t ts_start_;
+  uint64_t ts_end_;
+};
+
+// Operation that only returns a Status.
+// Example operations: DB::Write(), Iterator::Seek() and
+// Iterator::SeekForPrev().
+class StatusOnlyTraceExecutionResult : public TraceExecutionResult {
+ public:
+  StatusOnlyTraceExecutionResult(Status status, uint64_t start_timestamp,
+                                 uint64_t end_timestamp, TraceType trace_type);
+
+  virtual ~StatusOnlyTraceExecutionResult() override = default;
+
+  // Return value of DB::Write(), etc.
+  virtual Status GetStatus() const;
+
+  virtual Status Accept(Handler* handler) override;
+
+ private:
+  Status status_;
+};
+
+// Operation that returns a Status and a value.
+// Example operation: DB::Get()
+class SingleValueTraceExecutionResult : public TraceExecutionResult {
+ public:
+  SingleValueTraceExecutionResult(Status status, const std::string& value,
+                                  uint64_t start_timestamp,
+                                  uint64_t end_timestamp, TraceType trace_type);
+
+  SingleValueTraceExecutionResult(Status status, std::string&& value,
+                                  uint64_t start_timestamp,
+                                  uint64_t end_timestamp, TraceType trace_type);
+
+  virtual ~SingleValueTraceExecutionResult() override;
+
+  // Return status of DB::Get(), etc.
+  virtual Status GetStatus() const;
+
+  // Value for the searched key.
+  virtual std::string GetValue() const;
+
+  virtual Status Accept(Handler* handler) override;
+
+ private:
+  Status status_;
+  std::string value_;
+};
+
+// Operation that returns a vector of values and status.
+// Example operation: DB::MultiGet()
+class MultiValuesTraceExecutionResult : public TraceExecutionResult {
+ public:
+  MultiValuesTraceExecutionResult(const std::vector<Status>& multi_status,
+                                  const std::vector<std::string>& values,
+                                  uint64_t start_timestamp,
+                                  uint64_t end_timestamp, TraceType trace_type);
+
+  MultiValuesTraceExecutionResult(std::vector<Status>&& multi_status,
+                                  std::vector<std::string>&& values,
+                                  uint64_t start_timestamp,
+                                  uint64_t end_timestamp, TraceType trace_type);
+
+  virtual ~MultiValuesTraceExecutionResult() override;
+
+  // Return status of DB::MultiGet(), etc.
+  virtual std::vector<Status> GetMultiStatus() const;
+
+  // Values for the searched keys.
+  virtual std::vector<std::string> GetValues() const;
```

Comment:
Sorry had some questions on these interfaces too.

- I find it hard to imagine the typical use case requires the vector returned by `GetValues()` outlives the MultiValuesTraceExecutionResult. Can we return a const reference to the member variable instead?
- I also find it hard to imagine the `std::vector` move constructor consumes meaningful time that makes two `MultiValuesTraceExecutionResult`s worthwhile. `std::string` was a bit controversial because of the short string optimization that inlines a small buffer which had to be copied during a move. But `std::vector` just has size/capacity/pointer to buffer, it should be very very cheap to move. So I think just one constructor that takes vector by value (and callers can invoke it with a moved-from vector if they wish) is sufficient.

Reply:
Done